plugin/edge/concurrent_service_table.go

```diff
@@ -0,0 +1,79 @@
+package edge
+
+import (
+	"sync"
+)
+
+// ServiceTable specifies the mapping from service DNS names to edge sites.
+type ServiceTable map[string]Set
+
+// ServiceTableUpdate encapsulates all the information sent in a table update
+// from an edge site.
+type ServiceTableUpdate struct {
+	Meta  Site         `json:"meta"`
+	Event ServiceEvent `json:"event"`
+}
+
+// ConcurrentServiceTable is a table that can be safely shared between goroutines.
+type ConcurrentServiceTable struct {
+	sync.RWMutex
+	table ServiceTable
+}
+
+// NewConcurrentServiceTable creates a new concurrent table.
+func NewConcurrentServiceTable() *ConcurrentServiceTable {
+	return &ConcurrentServiceTable{
+		table: make(ServiceTable),
+	}
+}
+
+// Lookup performs a locked lookup for edge sites running a particular service.
+func (cst *ConcurrentServiceTable) Lookup(svc string) (Set, bool) {
+	cst.Lock()
+	defer cst.Unlock()
+	set, found := cst.table[svc]
+	return set, found
+}
+
+// Add adds a new entry to the table.
+func (cst *ConcurrentServiceTable) Add(meta Site, serviceName string) {
+
+	// Lock down the table.
+	cst.Lock()
+	defer cst.Unlock()
+
+	// Add the new site.
+	if edgeSites, found := cst.table[serviceName]; found {
+		edgeSites.Add(meta)
+	} else {
+		newSet := NewSet()
+		newSet.Add(meta)
+		cst.table[serviceName] = newSet
+	}
+
+	// Log the new table.
+	if svcDebugMode {
+		log.Infof("Updated table: %+v", cst.table)
```

Comment:
This log message is pretty noisy on `central`. Just thinking about scale here -- spamming the entire table to stdout best?

Reply:
I suppose we could introduce another level of debug logging. I think it's easier to see the current state of the table within a couple lines and use Cmd+F to find what you're looking for, as opposed to scrolling up and trying to look for each individual event that came through over potentially a very long time.

Reply:
I saw you imported a logger, that should control log levels without the `if`